File: Algorithms/JavaScript/MergeSort.js
//----------------------Iterative Approach--------------------------

function merge(arr1, arr2) {
    // Make a new array, and 2 pointers to keep track of elements of arr1 and arr2     
    let res = [],
        i = 0,
        j = 0;
    
    // Loop until either arr1 or arr2 becomes empty
    while (i < arr1.length && j < arr2.length) {
        // If the current element of arr1 is lesser than that of arr2, push arr1[i] and increment i         
        if (arr1[i] < arr2[j]) {
            res.push(arr1[i]);
            i++;
        } else {
            res.push(arr2[j]);
            j++;
        }
    }

    // Add the rest of the remining subarray, to our new array
    while (i < arr1.length) {
        res.push(arr1[i]);
        i++;
    }
    while (j < arr2.length) {
        res.push(arr2[j]);
        j++;
    }
    return res;
}


//----------------------Recursive Approach--------------------------
function mergeSort(arr) {
    // Base case
    if (arr.length <= 1) return arr;
  
    // Splitting into two halves
    let mid = Math.floor(arr.length / 2);
    let left = mergeSort(arr.slice(0, mid));
    let right = mergeSort(arr.slice(mid));
  
    // merging the two sorted halves
    return merge(left, right);
}

File: Algorithms/Java/merge_sort.java
import java.util.*;
public class merge_sort 
{
    public void Merge(int a[],int s,int mid,int f)
    {
        int l1=mid-s+1;
        int l2=f-mid;
        int left[]=new int[l1];
        int right[]=new int[l2];
        int i=0;
        int l=0;int j=0;
        for(i=0;i<l1;i++)
        left[i]=a[s+i];//fill left array

        for(i=0;i<l2;i++)
        right[i]=a[mid+1+i];//fill the right array
        
        i=s;
        while(l<l1 && j<l2)
        {
            if(left[l]<right[j])
            {
                a[i]=left[l];
                l++;
            }
            else
            {
                a[i]=right[j];
                j++;
            }
            i++;
        }
        while (l < l1)
        {
            a[i] = left[l];
            i++;
            l++;
        }
        while (j < l2)
        {
            a[i] = right[j];
            j++;
            i++;
        }
    }
    public void MergeSort(int a[],int s,int f)
    {
        if(s<f)
        {
            int mid=(s+f)/2;
            MergeSort(a, s, mid);
            MergeSort(a, mid+1, f);
            Merge(a,s,mid,f);
        }
    }
    public static void main(String args[])
    {
        Scanner sc=new Scanner(System.in);
        System.out.println("Input size of array");
        int n=sc.nextInt();
        int a[]=new int[n];
        System.out.println("Input "+n+" elements to fill the array");
        for(int i=0;i<n;i++)
        a[i]=sc.nextInt();
        merge_sort ob=new merge_sort();
        ob.MergeSort(a,0,n-1);
        System.out.println("Sorted Array is:");
        for(int i=0;i<n;i++)
        System.out.print(a[i]+" ");
    }    
}
